test/jasmine/tests/hover_label_test.js

```diff
@@ -501,6 +501,128 @@ describe('hover info', function() {
         });
     });
 
+    describe('hover info for negative data on a log axis', function() {
+        it('shows negative data even though it is infinitely off-screen', function(done) {
+            var gd = createGraphDiv();
+
+            Plotly.plot(gd, [{x: [1, 2, 3], y: [1, -5, 10]}], {
+                yaxis: {type: 'log'},
+                width: 500,
+                height: 400,
+                margin: {l: 0, t: 0, r: 0, b: 0}
+            })
+            .then(function() {
+                _hover(gd, 250, 200);
+                assertHoverLabelContent({
+                    nums: '\u22125', // unicode minus
+                    axis: '2'
+                });
+            })
+            .catch(fail)
+            .then(done);
+        });
+    });
+
+    describe('histogram hover info', function() {
+        it('shows the data range when bins have multiple values', function(done) {
+            var gd = createGraphDiv();
+
+            Plotly.plot(gd, [{
+                x: [0, 2, 3, 4, 5, 6, 7],
+                xbins: {start: -0.5, end: 8.5, size: 3},
+                type: 'histogram'
+            }], {
+                width: 500,
+                height: 400,
+                margin: {l: 0, t: 0, r: 0, b: 0}
+            })
+            .then(function() {
+                _hover(gd, 250, 200);
+                assertHoverLabelContent({
+                    nums: '3',
+                    axis: '3 - 5'
+                });
+            })
+            .catch(fail)
+            .then(done);
+        });
+
+        it('shows the exact data when bins have single values', function(done) {
+            var gd = createGraphDiv();
+
+            Plotly.plot(gd, [{
+                // even though the data aren't regularly spaced, each bin only has
+                // one data value in it so we see exactly that value
+                x: [0, 0, 3.3, 3.3, 3.3, 7, 7],
+                xbins: {start: -0.5, end: 8.5, size: 3},
+                type: 'histogram'
+            }], {
+                width: 500,
+                height: 400,
+                margin: {l: 0, t: 0, r: 0, b: 0}
+            })
+            .then(function() {
+                _hover(gd, 250, 200);
+                assertHoverLabelContent({
+                    nums: '3',
+                    axis: '3.3'
```

Comment:
⚠️  another potentially controversial decision: whenever *every* bin contains a single unique value, label exactly that value, not a range, even if those single values are not at the same places within the bins. But as soon as you have any bin with two distinct values in it, we use the range logic.

Note: for histogram2d I applied this independently to each axis, not to each brick. So if one x brick has only the point (2, 3) and one above it has only the point (2.1, 4), this will make the x bins show ranges, because *that one x bin* has multiple values, even though *each brick* has unique values in it.

Reply:
Nice touch here. I was worried that category histograms would get e.g, [`apples - lemons`](http://rickyreusser.com/plotly-mock-viewer/#hist2d_summed) bin hover labels. :ok_hand:

Reply:
That *can* happen - but only if the bins *actually* contain apples and lemons.
```js
Plotly.newPlot(gd, [{
  x: ['apples','apples','apples','lemons','lemons','peaches','peaches','pears'],
  type: 'histogram',
  xbins: {start: -0.5, end: 3.5, size: 2}
}],{
  width: 400,
  height: 400
})
```

![screen shot 2017-10-24 at 11 43 19 am](https://user-images.githubusercontent.com/2678795/31953198-9526ac18-b8b0-11e7-9871-4e758570f02b.png)

It's possible to put as many categories as you want into one bin, and as with numbers we'll only show the first and last ones in the label. Never happens automatically since it's in general confusing, but if you want to do it we won't stop you!

Reply:
> but if you want to do it we won't stop you!

Cool. Would you mind adding a test case for this situation?

Reply:
> Cool. Would you mind adding a test case for this situation?

https://github.com/plotly/plotly.js/pull/2113/commits/0331a16845be937de4cedac301979a0139249fa1